db/virtual_tables.cc

```diff
@@ -975,12 +983,185 @@ class raft_state_table : public streaming_virtual_table {
     }
 };
 
+/// Base class for virtual tables which are supposed to serve data from group0 leader.
+/// If queried on a different node, read will be redirected to the loader.
+/// Implementations need to override execute_on_leader(), which will be executed
+/// on shard0 of current group0 leader.
+///
+/// Current implementation is suitable for tables which are relatively small as all
+/// data is materialized in memory and queried in one page.
+class group0_virtual_table : public memtable_filling_virtual_table {
+private:
+    sharded<service::raft_group_registry>& _raft_gr;
+    sharded<netw::messaging_service>& _ms;
+public:
+    group0_virtual_table(schema_ptr s,
+                         sharded<service::raft_group_registry>& raft_gr,
+                         sharded<netw::messaging_service>& ms)
+        : memtable_filling_virtual_table(s)
+        , _raft_gr(raft_gr)
+        , _ms(ms) {
+    }
+
+    future<raft::server_id> get_leader(const reader_permit& permit) const {
+        while (!_raft_gr.local().group0().current_leader()) {
+            abort_on_expiry aoe(permit.timeout());
+            reader_permit::awaits_guard ag(permit);
+            co_await _raft_gr.local().group0().read_barrier(&aoe.abort_source());
+        }
+
+        co_return _raft_gr.local().group0().current_leader();
+    }
+
+    future<bool> is_leader(const reader_permit& permit) const {
+        auto leader = co_await get_leader(permit);
+        co_return leader == _raft_gr.local().get_my_raft_id();
+    }
+
+    future<> redirect_to_leader(std::function<void(mutation)> mutation_sink, reader_permit permit) {
+        auto leader = co_await get_leader(permit);
+        auto cmd = query::read_command(_s->id(),
+                                       _s->version(),
+                                       partition_slice_builder(*_s).build(),
+                                       query::max_result_size(256*1024*1024), // Sanity limit
+                                       query::tombstone_limit::max,
+                                       query::row_limit(query::max_rows),
+                                       query::partition_limit(query::max_partitions));
+        reader_permit::awaits_guard ag(permit);
+        auto&& [result, hit_rate, opt_exception] = co_await ser::storage_proxy_rpc_verbs::send_read_mutation_data(&_ms.local(),
+              locator::host_id(leader.uuid()),
+              permit.timeout(),
+              cmd,
+              query::full_partition_range,
+              {});
+        if (opt_exception.has_value() && *opt_exception) {
+            co_await coroutine::return_exception_ptr((*opt_exception).into_exception_ptr());
+        }
+        for (auto&& p : result.partitions()) {
+            mutation_sink(p.mut().unfreeze(_s));
```

Comment:
maybe_yield? at this level we don't know the partition count.

Reply:
Yes